misk-hibernate-testing/src/main/kotlin/misk/jdbc/MySqlScaleSafetyChecks.kt

```diff
@@ -0,0 +1,117 @@
+package misk.jdbc
+
+import com.squareup.moshi.Moshi
+import com.zaxxer.hikari.util.DriverDataSource
+import misk.environment.Environment
+import misk.hibernate.Check
+import misk.hibernate.Transacter
+import misk.vitess.StartDatabaseService
+import mu.KotlinLogging
+import net.ttddyy.dsproxy.proxy.ProxyConfig
+import net.ttddyy.dsproxy.support.ProxyDataSource
+import okhttp3.OkHttpClient
+import java.sql.Connection
+import java.sql.SQLException
+import java.sql.SQLSyntaxErrorException
+import java.sql.Timestamp
+import java.util.Properties
+import javax.inject.Inject
+import javax.inject.Singleton
+import javax.sql.DataSource
+
+@Singleton
+class MySqlScaleSafetyChecks @Inject constructor(
+  override val config: DataSourceConfig,
+  override val okHttpClient: OkHttpClient,
+  override val moshi: Moshi,
+  override val transacter: Transacter
+) : AbstractScaleSafetyChecks(config, okHttpClient, moshi, transacter) {
+  private val fullTableScanDetector = TableScanDetector()
+  private val connection: Connection by lazy { connect()!! }
+
+  override fun decorate(dataSource: DataSource): DataSource {
+    val proxy = ProxyDataSource(dataSource)
+
+    connection.createStatement().use { statement ->
+      statement.execute("SET GLOBAL log_output = 'TABLE'")
+      statement.execute("SET GLOBAL general_log = 1")
+    }
+
+    proxy.proxyConfig = ProxyConfig.Builder()
+      .methodListener(fullTableScanDetector)
+      .build()
+    proxy.addListener(fullTableScanDetector)
+    return proxy
+  }
+
+  override fun connect(): Connection? {
+    return try {
+      DriverDataSource(
+        config.buildJdbcUrl(Environment.TESTING),
+        config.type.driverClassName,
+        Properties(),
+        config.username,
+        config.password
+      ).connection
+    } catch (e: SQLException) {
+      throw IllegalStateException("Could not connect to test MySQL server!", e)
+    }
+  }
+
+  inner class TableScanDetector : ExtendedQueryExecutionListener() {
+    private val mysqlTimeBeforeQuery: ThreadLocal<Timestamp?> =
+      ThreadLocal.withInitial { null }
+
+    override fun beforeQuery(query: String) {
+      if (!transacter.isCheckEnabled(Check.TABLE_SCAN)) return
+
+      mysqlTimeBeforeQuery.set(connection.createStatement().use { statement ->
+        statement.executeQuery("SELECT MAX(event_time) FROM mysql.general_log")
+          .map { it.getTimestamp(1) }
+          .singleOrNull()
+      })
+    }
+
+    override fun afterQuery(query: String) {
+      if (!transacter.isCheckEnabled(Check.TABLE_SCAN)) return
+      val mysqlTime = mysqlTimeBeforeQuery.get() ?: return
+      val queries = extractQueriesSince(mysqlTime)
+
+      for (rawQuery in queries) {
```

Comment:
I think the rest of this method could be reused between the Vitess and the MySQL checkers. Possibly just as static methods.